cmd/snapshot-controller/main.go

```diff
@@ -59,6 +63,40 @@ var (
 	version = "unknown"
 )
 
+// Checks that the VolumeSnapshot v1 CRDs.
+func ensureCustomResourceDefinitionsExist(client *clientset.Clientset) error {
+	timeoutCh := make(chan struct{})
+	go func() {
+		defer close(timeoutCh)
+		time.Sleep(10 * time.Second)
+	}()
+
+	condition := func() (bool, error) {
+		var err error
+		_, err = client.SnapshotV1().VolumeSnapshots("kube-system").List(context.TODO(), metav1.ListOptions{})
```

Comment:
`VolumeSnapshots` needs a namespace, I was not sure if this should be kube-system although I'm not sure it exists in any cluster or if I should just check for the VolumeSnapshotClasses / VolumeSnapshotContents

Reply:
We can't assume VolumeSnapshots are in "kube-system" namespace.

Also I see that you are checking if there are any VolumeSnapshot/VolumeSnapshotContent/VolumeSnapshotClass resources.  Can you just check if the CRDs are deployed?

Reply:
The purpose of the check is that the endpoint exists, it doesn't matter if there's nothing in the namespace. 

I was also thinking of checking the CRD directly but that will require new rbacs.

Reply:
If the response of the list function, not only caused  by the crd not exists, maybe the problem of the network can effect it.

Reply:
Can you check if the specific error indicates CRDs are not deployed?

Reply:
![image](https://user-images.githubusercontent.com/13997135/115653261-6d8c5280-a361-11eb-80ba-8f1fe21827b4.png)
 
demo for crd.

Reply:
I believe both calls are the same, I did the following experiment with `volumesnapshotclasses` 

The dynamic client is returning unstructured data of the results of GET volumesnapshots (not of the CRD), I tried to make a request with kube-proxy and curl:

- with the dynamic client we're doing a `GET /apis/GROUP/VERSION/RESOURCE/RESOURCENAME`

```
# v1beta1
curl localhost:8080/apis/snapshot.storage.k8s.io/v1beta1/volumesnapshotclasses/
{"apiVersion":"snapshot.storage.k8s.io/v1beta1","items":[],"kind":"VolumeSnapshotClassList","metadata":{"continue":"","resourceVersion":"58636","selfLink":"/apis/snapshot.storage.k8s.io/v1beta1/volumesnapshotclasses/"}}

# v1
curl localhost:8080/apis/snapshot.storage.k8s.io/v1/volumesnapshotclasses/
404 page not found
```

- with the snapshot v1 client and kubectl I saw the following

```
# v1beta1
kubectl get volumesnapshotclasses.v1beta1.snapshot.storage.k8s.io --v=7
I0422 08:06:43.695353   77906 loader.go:375] Config loaded from file:  bmctl-workspace/ci-6ecffb9bde2ccf8/ci-6ecffb9bde2ccf8-kubeconfig
I0422 08:06:43.705084   77906 round_trippers.go:421] GET https://10.200.0.5:443/apis/snapshot.storage.k8s.io/v1beta1/volumesnapshotclasses?limit=500
I0422 08:06:43.705120   77906 round_trippers.go:428] Request Headers:
I0422 08:06:43.705131   77906 round_trippers.go:432]     Accept: application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json
I0422 08:06:43.705151   77906 round_trippers.go:432]     User-Agent: kubectl/v1.19.7 (linux/amd64) kubernetes/1dd5338
I0422 08:06:43.714630   77906 round_trippers.go:447] Response Status: 200 OK in 9 milliseconds
No resources found

#v1
kubectl get volumesnapshotclasses.v1.snapshot.storage.k8s.io --v=6
I0422 08:18:48.320835   83848 loader.go:375] Config loaded from file:  bmctl-workspace/ci-6ecffb9bde2ccf8/ci-6ecffb9bde2ccf8-kubeconfig
I0422 08:18:48.329195   83848 discovery.go:214] Invalidating discovery information
I0422 08:18:48.338454   83848 round_trippers.go:444] GET https://10.200.0.5:443/api?timeout=32s 200 OK in 9 milliseconds
I0422 08:18:48.348421   83848 round_trippers.go:444] GET https://10.200.0.5:443/apis?timeout=32s 200 OK in 1 milliseconds
... lots of OKs ...
F0422 08:18:48.792106   83848 helpers.go:115] error: the server doesn't have a resource type "volumesnapshotclasses"
```

The endpoints for both the dynamic client and the snapshot v1 client are the same, if our objective is to check the CRDs I think I could use another client like the apiextensions client to get the CRDs, check that they have the v1 version and check that the CRD has the `Established` condition as the docs say e.g.

```
kubectl get crd volumesnapshots.snapshot.storage.k8s.io --v=7 -o custom-columns=NAME:.metadata.name,VERSIONS:.spec.versions.*.name,CONDITIONS:.status.conditions.*.type
I0422 08:17:25.251236   83018 loader.go:375] Config loaded from file:  bmctl-workspace/ci-6ecffb9bde2ccf8/ci-6ecffb9bde2ccf8-kubeconfig
I0422 08:17:25.260059   83018 round_trippers.go:421] GET https://10.200.0.5:443/apis/apiextensions.k8s.io/v1/customresourcedefinitions/volumesnapshots.snapshot.storage.k8s.io
I0422 08:17:25.260089   83018 round_trippers.go:428] Request Headers:
I0422 08:17:25.260096   83018 round_trippers.go:432]     Accept: application/json
I0422 08:17:25.260101   83018 round_trippers.go:432]     User-Agent: kubectl/v1.19.7 (linux/amd64) kubernetes/1dd5338
I0422 08:17:25.273409   83018 round_trippers.go:447] Response Status: 200 OK in 13 milliseconds
I0422 08:17:25.274596   83018 table_printer.go:45] Unable to decode server response into a Table. Falling back to hardcoded types: attempt to decode non-Table object
NAME                                      VERSIONS   CONDITIONS
volumesnapshots.snapshot.storage.k8s.io   v1beta1    KubernetesAPIApprovalPolicyConformant,NamesAccepted,Established
```

Thanks for the comments! This was a really useful exercise to know more about how the API server works

Reply:
> Can you check if the specific error indicates CRDs are not deployed?

I'm not sure it's important to check that specific error. If we're not able to access the API server due to network issues, that's also a problem (although may be temporary)

Reply:
As @msau42 mentioned this change is doable however we'd need to add new rules to the RBAC e.g.

```
- apiGroups:
  - apiextensions.k8s.io
  resources:
  - customresourcedefinitions
  verbs:
  - get
```

I didn't try this in a cluster yet but this would be an approach:

```
// Checks that the VolumeSnapshot v1 CRDs exist and that they have the Established condition.
func ensureCRDsExist(client apiextensionsclientset.Interface) error {
	crdHasV1Version := func(crd *apiextensionsv1.CustomResourceDefinition) bool {
		for _, version := range crd.Spec.Versions {
			if version.Name == snapshotv1.SchemeGroupVersion.Version {
				return true
			}
		}
		return false
	}

	crdHasEstablishedCondition := func(crd *apiextensionsv1.CustomResourceDefinition) bool {
		for _, condition := range crd.Status.Conditions {
			if condition.Type == apiextensionsv1.Established {
				return true
			}
		}
		return false
	}

	condition := func() (bool, error) {
		groupResources := []string{
			snapshotv1.Resource("volumesnapshots").String(),
			snapshotv1.Resource("volumesnapshotclasses").String(),
			snapshotv1.Resource("volumesnapshotcontents").String(),
		}
		for _, groupResource := range groupResources {
			crd, err := client.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), groupResource, metav1.GetOptions{})
			if err != nil {
				klog.Errorf("Failed to get CRD %s with error: %+v", groupResource, err)
				return false, nil
			}
			// Check that the CRD has the v1 version and that it has the Established condition
			// The Established condition was mentioned in https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition
			if !crdHasV1Version(crd) {
				klog.Errorf("CRD %s doesn't declare a v1 version", groupResource)
				return false, nil
			}
			if !crdHasEstablishedCondition(crd) {
				klog.Errorf("CRD %s doesn't have an Established condition", groupResource)
				return false, nil
			}
		}
		return true, nil
	}

	// with a Factor of 1.5 we wait up to 7.5 seconds (the 10th attempt)
	backoff := wait.Backoff{
		Duration: 100 * time.Millisecond,
		Factor:   1.5,
		Steps:    10,
	}
	if err := wait.ExponentialBackoff(backoff, condition); err != nil {
		return err
	}
	return nil
}
```

I believe that just by testing that the endpoint exist with the snapshotclient might give us a good indication that the CRD had an Established condition.

Reply:
> `VolumeSnapshots` needs a namespace, I was not sure if this should be kube-system although I'm not sure it exists in any cluster or if I should just check for the VolumeSnapshotClasses / VolumeSnapshotContents

if we deploy the sidecar under othe namespace, it can list the resoures under the kube-system?

Reply:
thanks for pointing this out, I created https://github.com/kubernetes-csi/external-snapshotter/issues/514 to fix this